compiler/resolution/functionResolution.cpp

```diff
@@ -4297,32 +4297,41 @@ static void generateUnresolvedMsg(CallInfo& info, Vec<FnSymbol*>& visibleFns) {
     sortExampleCandidates(info, visibleFns);
 
     int nPrintDetails = 1;
-    int nPrint = 3;
+    int nPrint = fPrintAllCandidates ? INT_MAX : 3;
 
 
-    // If there is more than one visible function and we haven't been
-    // asked to print all candidates, let's try to eliminate "worse"
-    // matches (e.g., method call / standalone function call
-    // mismatches).
     Vec<FnSymbol*> filteredFns;
-    if (visibleFns.n > 1 && fPrintAllCandidates == false) {
-      forv_Vec(FnSymbol, fn, visibleFns) {
-        if (!obviousMismatch(call, fn)) {
-          filteredFns.add(fn);
+    if (fPrintAllCandidates) {
+      filteredFns = visibleFns;
+    } else {
+      // If there is more than one visible function and we haven't been
+      // asked to print all candidates, let's try to eliminate "worse"
+      // matches (e.g., method call / standalone function call
+      // mismatches).
+      if (visibleFns.n > 1) {
```

Comment:
Maybe the issue is that you meant for this condition to be combined with 4304? The way it is now, I think it would put nothing in filteredFns if there is 1 candidate, and then in the logic below, add all of the visible functions.

Anyway I think it would be nicer to put the logic about filteredFns.n == 0 in this else block rather than in another condition below.

I'm thinking

```
 if visibleFns.n <= 1 || fPrintAllCandidates
   filteredFns = visibleFns;
 else
   add not-obvious-mismatches to filteredFns
   if filteredFns is empty, copy over the first nPrint
 ```
 
 but of course there are lots of other ways to write it

Reply:
I take that back, I don't think that's the issue, but you can adjust the logic to my suggestion if you like it better.

Reply:
It was more like an off-by-one error, but fixing it gave me the motivation to do some other cleanups that had been in the back of m mind.  I also ended up doing something like your suggestion here, which is a bit more 6 vs. 1/2 dozen to me but definitely makes the n==1 case less convoluted...  Re-testing now.

Reply:
@mppf: This is back to testing cleanly and is fairly improved in terms of its refactoring, I think.  Would you mind taking another look to see if you agree before I merge it?  Thanks!